template/template.go

```diff
@@ -31,28 +31,77 @@ type Template struct {
 	locale   language.Tag
 }
 
-type variable struct {
-	name        *string
-	formatValue *string // TODO(mvilks): "selectValue" to be implemented
-	expression  ast.Expression
+type Result struct {
+	value     any
+	selectKey func(keys []string) string
+	format    func() string
+	err       error
 }
 
-func newVariable(e ast.Expression, name *string) variable { return variable{expression: e, name: name} }
+func (r *Result) Value() any {
+	if v, ok := r.value.(*Result); ok {
+		return v.Value()
+	}
+
+	return r.value
+}
 
-func (v variable) Format(e *executer) (string, error) {
-	if v.formatValue != nil {
-		return *v.formatValue, nil
+func (r *Result) String() string {
+	if r.format != nil {
+		return r.format()
 	}
 
-	val, err := e.resolveExpression(v.expression)
-	if err != nil && v.name != nil {
-		n := "{$" + *v.name + "}"
-		val = n
+	switch value := r.value.(type) {
+	default:
+		s, err := castAs[string](r.value) // if underlying type is not string, return error
+		if err != nil {
+			return "" // errorf("unsupported value type: %T: %w", r.value, err)
+		}
+
+		return s
+	case fmt.Stringer:
+		return value.String()
+	case string, []byte, []rune, int, int8, int16, int32, int64,
+		uint, uint8, uint16, uint32, uint64, float32, float64, bool,
+		complex64, complex128, error:
+		return fmt.Sprint(value)
 	}
+}
 
-	v.formatValue = &val
+func (r *Result) SelectKey(keys []string) string {
+	if r.selectKey != nil {
+		return r.selectKey(keys)
+	}
+
+	if r.format != nil {
+		return r.format()
+	}
```

Comment:
What happens when `format()` value does not match any key in keys?

Reply:
Also, if the struct has no format function, we can set a default function to select the key based on format value.

Reply:
The same applies to the format function. NewResult() sets default format() or selectKeys() if not set with `WithSelectKey` and `WithFormat` options.

Reply:
However, what does MF2 say about using a registry function for selection and formatting?

Reply:
If matching any key fails, the fallback catch-all key is returned.

Reply:
unicode-org/message-format-wg#852

Reply:
How strict do we want to be about formatting?

* We MUST return an error anyway if no formatting function is present
* Fall back to fallback value or value. Currently, your implementation falls back to value.

Reply:
We'll resolve it in separate PR.